datafusion/core/tests/user_defined/relation_planner.rs

```diff
@@ -0,0 +1,353 @@
+        .unwrap();
+
+    let results = collect_sql(&ctx, "SELECT * FROM custom_values").await;
+
+    let expected = "\
++---------+
+| column1 |
++---------+
+| 1       |
+| 2       |
+| 3       |
++---------+";
+    assert_eq!(batches_to_string(&results), expected);
+}
+
+#[tokio::test]
+async fn test_multiple_planners_first_wins() {
+    let ctx = SessionContext::new();
+
+    // Register multiple planners - first one wins
+    ctx.register_relation_planner(Arc::new(CustomValuesPlanner))
+        .unwrap();
+    ctx.register_relation_planner(Arc::new(StringTablePlanner))
+        .unwrap();
+
+    // CustomValuesPlanner handles this
+    let results = collect_sql(&ctx, "SELECT * FROM custom_values").await;
+    let expected = "\
++---------+
+| column1 |
++---------+
+| 1       |
+| 2       |
+| 3       |
++---------+";
+    assert_eq!(batches_to_string(&results), expected);
+
+    // StringTablePlanner handles this
+    let results = collect_sql(&ctx, "SELECT * FROM colors").await;
+    let expected = "\
++---------+
+| column1 |
++---------+
+| red     |
+| green   |
+| blue    |
++---------+";
+    assert_eq!(batches_to_string(&results), expected);
+}
+
+#[tokio::test]
+async fn test_planner_delegates_to_default() {
+    let ctx = SessionContext::new();
+
+    // Register a planner that always returns None
+    ctx.register_relation_planner(Arc::new(PassThroughPlanner))
+        .unwrap();
+
+    // Also register a real table
+    let schema = Arc::new(Schema::new(vec![Field::new(
+        "value",
+        DataType::Int64,
+        true,
+    )]));
+    let batch =
+        RecordBatch::try_new(schema.clone(), vec![Arc::new(Int64Array::from(vec![42]))])
+            .unwrap();
+    let table = MemTable::try_new(schema, vec![vec![batch]]).unwrap();
+    ctx.register_table("real_table", Arc::new(table)).unwrap();
+
+    // PassThroughPlanner returns None, so it delegates to the default planner
+    let results = collect_sql(&ctx, "SELECT * FROM real_table").await;
+    let expected = "\
++-------+
+| value |
++-------+
+| 42    |
++-------+";
+    assert_eq!(batches_to_string(&results), expected);
+}
+
+#[tokio::test]
+async fn test_planner_delegates_with_multiple_planners() {
+    let ctx = SessionContext::new();
+
+    // Register planners in order
+    ctx.register_relation_planner(Arc::new(PassThroughPlanner))
+        .unwrap();
+    ctx.register_relation_planner(Arc::new(CustomValuesPlanner))
+        .unwrap();
+    ctx.register_relation_planner(Arc::new(StringTablePlanner))
+        .unwrap();
+
+    // PassThroughPlanner returns None, CustomValuesPlanner handles it
+    let results = collect_sql(&ctx, "SELECT * FROM custom_values").await;
+    let expected = "\
++---------+
+| column1 |
++---------+
+| 1       |
+| 2       |
+| 3       |
++---------+";
+    assert_eq!(batches_to_string(&results), expected);
+
+    // PassThroughPlanner and CustomValuesPlanner both return None,
+    // StringTablePlanner handles it
```

Comment:
I think I start getting blurry idea. We go and get the table by their registered order? Why that is important?